node/core/pvf/src/prepare/worker.rs

```diff
@@ -0,0 +1,187 @@
+	/// This doesn't return an idle worker instance, thus this worker is no longer usable.
+	DidntMakeIt,
+}
+
+pub async fn start_work(
+	worker: IdleWorker,
+	code: Arc<Vec<u8>>,
+	artifact_path: PathBuf,
+	background_priority: bool,
+) -> Outcome {
+	let IdleWorker { mut stream, pid } = worker;
+
+	if background_priority {
+		renice(pid, NICENESS_BACKGROUND);
+	}
+
+	if let Err(err) = send_request(&mut stream, code).await {
+		tracing::warn!("failed to send a prepare request to pid={}: {:?}", pid, err);
+		return Outcome::DidntMakeIt;
+	}
+
+	// Wait for the result from the worker, keeping in mind that there may be a timeout, the
+	// worker may get killed, or something along these lines.
+	//
+	// In that case we should handle these gracefully by writing the artifact file by ourselves.
+	// We may potentially overwrite the artifact in rare cases where the worker didn't make
+	// it to report back the result.
+
+	enum Selected {
+		Done,
+		IoErr,
+		Deadline,
+	}
+
+	let selected = futures::select! {
+		artifact_path_bytes = framed_recv(&mut stream).fuse() => {
+			match artifact_path_bytes {
+				Ok(bytes) => {
+					if let Some(tmp_path) = bytes_to_path(&bytes) {
+						async_std::fs::rename(tmp_path, &artifact_path)
+							.await
+							.map(|_| Selected::Done)
+							.unwrap_or(Selected::IoErr)
+					} else {
+						Selected::IoErr
+					}
+				},
+				Err(_) => Selected::IoErr,
+			}
+		},
+		_ = Delay::new(COMPILATION_TIMEOUT).fuse() => Selected::Deadline,
+	};
+
+	match selected {
+		Selected::Done => {
+			renice(pid, NICENESS_FOREGROUND);
+			Outcome::Concluded(IdleWorker { stream, pid })
+		}
+		Selected::IoErr | Selected::Deadline => {
+			let bytes = Artifact::DidntMakeIt.serialize();
+			// best effort: there is nothing we can do here if the write fails.
+			let _ = async_std::fs::write(&artifact_path, &bytes).await;
+			Outcome::DidntMakeIt
+		}
+	}
+}
+
+async fn send_request(stream: &mut UnixStream, code: Arc<Vec<u8>>) -> io::Result<()> {
+	framed_send(stream, &*code).await?;
+	Ok(())
+}
+
+async fn recv_request(stream: &mut UnixStream) -> io::Result<Vec<u8>> {
+	let code = framed_recv(stream).await?;
+	Ok(code)
+}
+
+pub fn bump_priority(handle: &WorkerHandle) {
+	let pid = handle.id();
+	renice(pid, NICENESS_FOREGROUND);
+}
+
+fn renice(pid: u32, niceness: i32) {
+	// Consider upstreaming this to the `nix` crate.
+	unsafe {
+		if -1 == libc::setpriority(libc::PRIO_PROCESS, pid, niceness) {
+			let err = std::io::Error::last_os_error();
+			tracing::warn!(target: LOG_TARGET, "failed to set the priority: {:?}", err,);
+		}
+	}
+}
+
+/// The entrypoint that the spawned prepare worker should start with. The socket_path specifies
+/// the path to the socket used to communicate with the host.
+pub fn worker_entrypoint(socket_path: &str) {
+	worker_event_loop("prepare", socket_path, |mut stream| async move {
+		loop {
+			let code = recv_request(&mut stream).await?;
+
+			let artifact_bytes = prepare_artifact(&code).serialize();
```

Comment:
Maybe some logging here in the loop would be nice.

Reply:
Yeah. Initially I assumed there will be no logging for the worker processes, but turned out that the legacy executor enabled them.